controllers/kafkamanagement/kafkauser_controller.go

```diff
@@ -533,3 +551,432 @@ func (r *KafkaUserReconciler) getKafkaUserCredsFromSecret(
+		if err != nil {
+			l.Error(err, "Cannot create Kafka user mTLS certificate", "cert", newCert)
+			r.EventRecorder.Eventf(newObj, models.Warning, models.CreatingEvent,
+				"Cannot create user mTLS certificate. Reason: %v", err)
+		}
+
+		oldCerts = append(oldCerts, newCert)
+	}
+}
+
+func (r *KafkaUserReconciler) handleCreateCertificate(ctx context.Context, user *v1beta1.KafkaUser, l logr.Logger, certCreationData *v1beta1.CertificateCreationData) error {
+	username, _, err := r.getKafkaUserCredsFromSecret(user.Spec)
+	if err != nil {
+		l.Error(
+			err, "Cannot get Kafka user creds from secret",
+			"kafka user spec", user.Spec,
+		)
+		r.EventRecorder.Eventf(
+			user, models.Warning, models.FetchFailed,
+			"Fetch user credentials from secret is failed. Reason: %v",
+			err,
+		)
+
+		return err
+	}
+
+	if certCreationData.CSR == "" {
+		certCreationData.CSR, err = r.GenerateCSR(certCreationData)
+		if err != nil {
+			l.Error(err, "Cannot generate CSR for Kafka user certificate creation",
+				"user", user.Name,
+			)
+			r.EventRecorder.Eventf(
+				user, models.Warning, models.GenerateFailed,
+				"Generate CSR is failed. Reason: %v",
+				err,
+			)
+
+			return err
+		}
+	}
+
+	certResponse, err := r.API.CreateKafkaUserCertificate(instaclustr.KafkauserCertificatesEndpoint, certCreationData, username)
+	if err != nil {
+		l.Error(err, "Cannot create Kafka user mTLS certificate",
+			"user", user.Name,
+		)
+		r.EventRecorder.Eventf(
+			user, models.Warning, models.CreationFailed,
+			"Certificate creation is failed. Reason: %v",
+			err,
+		)
+
+		return err
+	}
+
+	newSecret := user.NewCertificateSecret(certCreationData.SecretName, certCreationData.SecretNamespace)
+	err = r.Client.Create(ctx, newSecret)
+	if err != nil {
+		if k8serrors.IsAlreadyExists(err) {
+			l.Error(err, "Cannot create Kafka user Cert Secret. Secret already exist",
+				"secret name", certCreationData.SecretName,
+			)
+			r.EventRecorder.Eventf(
+				user, models.Warning, models.CreationFailed,
+				"create user Cert Secret is failed. Reason: %v",
+				err,
+			)
+
+			return err
+		}
+		l.Error(err, "Cannot create Kafka user Cert Secret.",
+			"secret name", certCreationData.SecretName,
+		)
+		r.EventRecorder.Eventf(
+			user, models.Warning, models.CreationFailed,
+			"create user Cert Secret is failed. Reason: %v",
+			err,
+		)
+
+		return err
+
+	}
+
+	controllerutil.AddFinalizer(newSecret, models.DeletionFinalizer)
+
+	err = r.UpdateCertSecret(ctx, newSecret, certResponse)
+	if err != nil {
+		l.Error(err, "Cannot update certificate secret",
+			"user", user.Name,
+			"secret", newSecret.Name,
+		)
+		r.EventRecorder.Eventf(
+			user, models.Warning, models.UpdateFailed,
+			"Certificate secret update is failed. Reason: %v",
+			err,
+		)
+
+		return err
+	}
+
+	l.Info("Kafka user mTLS certificate has been created",
+		"User ID", user.GetID(certCreationData.ClusterID, username),
+	)
+
+	certCreationData.CSR = ""
+
+	return nil
+}
```

Comment:
Does certCreationData.CSR need to be empty in the end and why?

Reply:
refactored